### DISCUSSION TITLE
是否可以增加根据 process name 与 process id 规则去做路由分发？

### DISCUSSION THREAD

**Discussion by lemos1235** [eycorsican/leaf#370] [Ideas]
具体思路是根据 ip与 port获取相应的 TcpTable/UdpTable 对应的进程信息。目前看了一下 clash的 实现：

https://github.com/Dreamacro/clash/blob/d808576f98019529f098476df979c7bcf2822c84/component/process/process_windows.go#L32

**Comment by spongebob888** [comment]
感觉不难，但是要为全平台都添加的话，工作量应该也不小。

**Comment by eycorsican** [comment]
你要不要挑战一下？

**Reply by lemos1235** [reply]
提了一个 pr，用的第三方库实现的，我简单跑了一下 windows与 mac端，能用。

**Reply by eycorsican** [reply]
我在macOS上试了下你实现的，每次查找要花2~3秒，不是很理想。好像还阻塞了整个线程，没仔细看不确定

**Reply by spongebob888** [reply]
感觉可以将首次获得的进程信息存在session里。省去后续的系统调用

**Comment by lemos1235** [comment]
我发现一个问题，如果我使用非本机设备提供代理服务，速度是比较正常的。如果用本机的代理服务（clashX）或者本机使用了 proxfier，速度会很慢。
具体原因暂未弄明白。

```
[General]
dns-server = 223.5.5.5
tun = auto
loglevel = debug
always-real-ip = baidu

[Proxy]
Direct = direct
# 本机设备作代理服务器，速度很慢
SOCKS5 = socks,192.168.1.11,7890
# 另一台设备作代理服务器，速度正常
SOCKS5 = socks,192.168.1.7,1080

[Rule]
# will match process path containing `curl`
PROCESS-NAME, curl, SOCKS5
PROCESS-NAME, hey, SOCKS5
FINAL, Direct
```

日志如下
[本地代理服务-用时15s.txt](https://github.com/eycorsican/leaf/files/11154667/-.15s.txt)
[非本地代理服务-用时0.7s.txt](https://github.com/eycorsican/leaf/files/11154668/-.0.7s.txt)

**Reply by eycorsican** [reply]
> 我在macOS上试了下你实现的，每次查找要花2~3秒，不是很理想。好像还阻塞了整个线程，没仔细看不确定

看看是不是因为阻塞了

**Reply by lemos1235** [reply]
应该是 process_finder.rs 这个文件中的 let system = System::new_all() 这个调用会阻塞线程。

**Reply by lemos1235** [reply]
最新的提交，优化了一下代码，查找时间基本在100ms以内

**Reply by eycorsican** [reply]
我以前用Go做过这个功能，我记得在Windows上是30ms左右，macOS上是15ms左右，用Rust实现不应该比这个慢。。。

**Reply by eycorsican** [reply]
另外这种操作用 `spawn_blocking`(https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html) 放到单独线程做可能比较好

**Reply by lemos1235** [reply]
用 spawn_blocking 的话是需要将rule的所有 apply 方法改成 async 的么？我不是很熟悉这类用法。

**Comment by lemos1235** [comment]
emm，大概是提取进程信息用的算法不合适

eycorsican ***@***.***>于2023年4月8日 周六13:00写道：
